onnxscript/values.py

```diff
@@ -271,12 +299,93 @@ def __init__(
         self.source = source
         self.kwargs = kwargs
         self._param_schemas: Optional[tuple[ParamSchema, ...]] = None
+        self._opschema: Optional[onnx.defs.OpSchema] = None
 
     @property
     def name(self):
         """Returns the function name."""
         return self.opname
 
+    @property
+    def opschema(self) -> Optional[onnx.defs.OpSchema]:
+        """Construct an OpSchema from function_ir."""
+        if self._opschema is not None:
+            return self._opschema
+
+        if not _ONNX_OP_SCHEMA_WRITABLE:
+            return None
+
+        function_ir = self.function_ir
+        # Find all distinct types in the inputs and outputs
+        distinct_types = {arg.typeinfo for arg in function_ir.inputs}.union(
+            {arg.typeinfo for arg in function_ir.outputs}
+        )
+        # Create a mapping from type to a unique name
+        type_to_constraint = {}
+        for i, type_ in enumerate(distinct_types):
+            name = f"T{i}"
+            type_to_constraint[type_] = TypeConstraint(
+                name=type_annotation.get_type_constraint_name(type_) or name,
+                allowed_types=type_annotation.pytype_to_input_strings(type_),
+            )
+
+        formal_inputs = [
+            onnx.defs.OpSchema.FormalParameter(
+                arg.name,
+                type_to_constraint[arg.typeinfo].name,
+                param_option=onnx.defs.OpSchema.FormalParameterOption.Optional
+                if typing.get_origin(arg.typeinfo) is Union and typing.get_args(arg.typeinfo)
```

Comment:
May be worth abstracting this into a function and moving it into `type_annotation`. Doesn't it need to check if `None` is one the values in the union? We may also need to figure out how to distinguish the two kinds of optionals ONNX supports. (That is, extend the type-annotations to allow users to indicate which they intend.)

Reply:
Done. Thanks for catching the bug too

Reply:
> We may also need to figure out how to distinguish the two kinds of optionals ONNX supports. (That is, extend the type-annotations to allow users to indicate which they intend.)

Merging to the discussion below